domain-mapping/classes/class-dm-media.php

```diff
@@ -0,0 +1,393 @@
+	 * Used to map a URL to a Media domain. Any URL passed into this method **will** be mapped.
+	 *
+	 * @param  string $url URL to be modified.
+	 * @return string      URL with the domain changed to be from a Media domain.
+	 *
+	 * @since 2.2.0
+	 */
+	public function map_url( $url = '' ) {
+		/**
+		 * No Media domains or the URL is blank, then bail.
+		 */
+		if ( ! $this->can_map() || empty( $url ) ) {
+			return $url;
+		}
+
+		/**
+		 * Check to ensure the URL is on a domain we can map. This is also used to prevent double-mapping occurring.
+		 */
+		$do_map = false;
+		foreach ( $this->main_domains as $main_domain ) {
+			if ( false !== stripos( $url, $main_domain ) ) {
+				$do_map = true;
+				break;
+			}
+		}
+
+		if ( ! $do_map ) {
+			return $url;
+		}
+
+		/**
+		 * Alternate through the Media domains if there is more than one.
+		 */
+		$index = 0;
+
+		if ( $this->media_domains_count > 0 ) {
+			$index = wp_rand( 0, $this->media_domains_count );
+		}
+
+		$url = preg_replace(
+			"#://(" . implode( '|', $this->main_domains ) . ")#",
+			'://' . untrailingslashit( $this->media_domains[ $index ]->domain ),
+			$url
+		);
+
+		/**
+		 * Ensure the URL is HTTPS.
+		 */
+		return set_url_scheme( $url, 'https' );
+	}
+
+	/**
+	 * Apply filters to map / unmap asset domains on REST API.
+	 *
+	 * @return void
+	 *
+	 * @since 2.2.0
+	 */
+	public function prepare_rest() {
+		/**
+		 * Loop all post types with REST endpoints to fix the mapping for content.raw property.
+		 */
+		$rest_post_types = get_post_types( array( 'show_in_rest' => true ) );
+
+		foreach ( $rest_post_types as $post_type ) {
+			add_filter( "rest_prepare_{$post_type}", array( $this, 'prepare_rest_post_item' ), 10, 1 );
+		}
+	}
+
+	/**
+	 * Ensures the "raw" version of the content, typically used by Gutenberg through it's middleware pre-load / JS
+	 * hydrate process, gets handled the same as content (which runs through the `the_content` hook).
+	 *
+	 * @param  WP_REST_Response $item Individual post / item in the response that is being processed.
+	 * @return WP_REST_Response       Post / item with the content.raw, if present, mapped.
+	 *
+	 * @since 2.2.0
+	 */
+	public function prepare_rest_post_item( $item = null ) {
+		if ( isset( $item->data['content']['raw'] ) ) {
+			$item->data['content']['raw'] = $this->map( $item->data['content']['raw'] );
+		}
+
+		return $item;
+	}
+
+	/**
+	 * Used to unmap Media domains.
+	 *
+	 * @param  string $value Value that may contain Media domains.
+	 * @return string        Value with the Media domains removed and replaced with the unmapped domain.
+	 *
+	 * @since 2.2.0
+	 */
+	public function unmap( $value = '' ) {
+		if ( ! $this->can_map() || empty( $value ) ) {
+			return $value;
+		}
+
+		/**
+		 * Create an array of strings of the Media domains.
+		 */
+		$media_domains = wp_list_pluck( $this->media_domains, 'domain' );
+
+		/**
+		 * Replace the Media domains with the unmapped domain.
+		 */
+		return preg_replace(
```

Comment:
what happens if `$media_domains` is empty?

Reply:
In theory, that should not be possible. The earlier call to `$this->can_map()` ensures that `$this->media_domains` is populated:

https://github.com/cameronterry/dark-matter/pull/68/files/39a5aca3be7b8d5d1f14310b7a6a84ab6b98f75a#diff-55101bbc11342be7e784d9113dca1500fa1bb10930ec0c46c4a9f1b8e476d356R74-R83

That being said, I've added an extra check just in case there is an edge case I can't find / imagine here.